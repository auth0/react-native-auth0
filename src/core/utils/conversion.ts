/**
 * A private helper that converts a single snake_case or kebab-case string to camelCase.
 * e.g., 'user_profile' -> 'userProfile'
 */
export function snakeToCamel(str: string): string {
  var parts = str.split('_').filter((part) => part.length > 0);
  if (parts.length === 0) return '';

  return parts.reduce(function (p, c, index) {
    if (index === 0) {
      return c.charAt(0).toLowerCase() + c.slice(1);
    }
    return p + c.charAt(0).toUpperCase() + c.slice(1);
  }, '');
}

/**
 * Recursively traverses an object or an array and converts all keys from
 * snake_case to camelCase.
 *
 * @param data The object or array to be transformed.
 * @returns A new object or array with all keys in camelCase.
 */
export function deepCamelCase<T>(data: any): T {
  if (Array.isArray(data)) {
    return data.map((v) => deepCamelCase(v)) as T;
  }
  if (data && typeof data === 'object' && data.constructor === Object) {
    const newObj: { [key: string]: any } = {};
    for (const key in data) {
      if (Object.prototype.hasOwnProperty.call(data, key)) {
        const camelKey = snakeToCamel(key);
        newObj[camelKey] = deepCamelCase(data[key]);
      }
    }
    return newObj as T;
  }
  return data as T;
}

/**
 * Converts a JavaScript object into a URL-encoded query string.
 *
 * @remarks
 * Keys with `null` or `undefined` values are ignored.
 *
 * @example
 * ```
 * toUrlQueryParams({ a: 1, b: 'hello world' })
 * // returns "a=1&b=hello%20world"
 * ```
 *
 * @param params The object of parameters to convert.
 * @returns A URL-encoded query string.
 */
export function toUrlQueryParams(params: Record<string, any>): string {
  const searchParams = new URLSearchParams();
  for (const key in params) {
    if (Object.prototype.hasOwnProperty.call(params, key)) {
      const value = params[key];
      if (value !== null && value !== undefined) {
        searchParams.append(key, String(value));
      }
    }
  }
  return searchParams.toString();
}
