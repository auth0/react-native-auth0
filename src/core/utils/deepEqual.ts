export function deepEqual<T>(x: T, y: T): boolean {
  if (x === y) {
    return true;
  } else if (
    typeof x == 'object' &&
    x != null &&
    typeof y == 'object' &&
    y != null
  ) {
    if (Object.keys(x).length != Object.keys(y).length) return false;

    for (const prop in x) {
      if (Object.prototype.hasOwnProperty.call(y, prop)) {
        if (!deepEqual(x[prop], y[prop])) return false;
      } else return false;
    }
    return true;
  } else {
    return false;
  }
}
