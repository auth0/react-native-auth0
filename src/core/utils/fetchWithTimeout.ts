import { AuthError } from '../models';

export class TimeoutError extends AuthError {
  constructor(message: string) {
    super('TimeoutError', message, { code: 'timeout' });
  }
}

export function fetchWithTimeout(
  url: RequestInfo,
  options: RequestInit,
  timeoutMs: number
): Promise<Response> {
  const controller = new AbortController();
  const signal = controller.signal;
  options.signal = signal;

  const timeout = setTimeout(() => controller.abort(), timeoutMs);

  return fetch(url, options)
    .catch((err) => {
      if (err.name === 'AbortError') {
        throw new TimeoutError(`Request timed out after ${timeoutMs}ms`);
      }
      throw err;
    })
    .finally(() => {
      clearTimeout(timeout);
    });
}
