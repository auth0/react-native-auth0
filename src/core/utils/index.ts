export * from './validation';
export * from './conversion';
export * from './fetchWithTimeout';
export * from './scope';
