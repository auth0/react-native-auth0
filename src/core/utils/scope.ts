/**
 * Ensures that the 'openid' scope is always present, which is required
 * for receiving an ID token. It also adds 'profile' and 'email' as
 * sensible defaults if no other scopes are requested.
 *
 * @param inputScopes A space-separated string of scopes provided by the user.
 * @returns A finalized, space-separated string of scopes.
 */
export function finalizeScope(inputScopes?: string): string {
  const defaultScopes = ['openid', 'profile', 'email'];

  if (!inputScopes?.trim()) {
    return defaultScopes.join(' ');
  }

  const providedScopes = new Set(inputScopes.split(' ').filter((s) => s));

  // Ensure 'openid' is always included.
  providedScopes.add('openid');

  return Array.from(providedScopes).join(' ');
}
