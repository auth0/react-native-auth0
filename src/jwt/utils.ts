export const idTokenNonProfileClaims = new Set([
  'iss',
  'aud',
  'exp',
  'nbf',
  'iat',
  'jti',
  'azp',
  'nonce',
  'auth_time',
  'at_hash',
  'c_hash',
  'acr',
  'amr',
  'sub_jwk',
  'cnf',
  'sip_from_tag',
  'sip_date',
  'sip_callid',
  'sip_cseq_num',
  'sip_via_branch',
  'orig',
  'dest',
  'mky',
  'events',
  'toe',
  'txn',
  'rph',
  'sid',
  'vot',
  'vtm',
]);
