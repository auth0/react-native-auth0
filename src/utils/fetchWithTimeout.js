import BaseError from './baseError';

export class TimeoutError extends BaseError {
  constructor(msg) {
    super('TimeoutError', msg);
  }
}

function makeTimeout(timeoutMs) {
  let timerId = null;

  const promise = new Promise((_, reject) => {
    timerId = setTimeout(() => {
      reject(new TimeoutError('Timeout'));
    }, timeoutMs);
  });

  return {
    timerId: timerId,
    promise: promise,
  };
}

export function fetchWithTimeout(url, options, timeoutMs) {
  const {promise: timeoutPromise, timerId} = makeTimeout(timeoutMs);
  const abortController = new AbortController();

  return Promise.race([
    fetch(url, {
      ...options,
      signal: abortController.signal,
    }),
    timeoutPromise,
  ])
    .catch(error => {
      if (error instanceof TimeoutError) {
        abortController.abort();
      }

      throw error;
    })
    .then(response => {
      clearTimeout(timerId);
      return response;
    });
}
