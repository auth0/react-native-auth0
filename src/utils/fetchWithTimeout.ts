import BaseError from './baseError';

export class TimeoutError extends BaseError {
  constructor(msg: string) {
    super('TimeoutError', msg);
  }
}

function makeTimeout(timeoutMs: number) {
  let timerId: NodeJS.Timeout | null = null;

  const promise = new Promise((_, reject) => {
    timerId = setTimeout(() => {
      reject(new TimeoutError('Timeout'));
    }, timeoutMs);
  });

  return {
    timerId: timerId,
    promise: promise,
  };
}

/**
 * @private
 */
export function fetchWithTimeout(
  url: string,
  options: RequestInit,
  timeoutMs: number,
): Promise<Response> {
  const {
    promise: timeoutPromise,
    timerId,
  }: {promise: Promise<unknown>; timerId: any} = makeTimeout(timeoutMs);
  const abortController = new AbortController();

  return Promise.race([
    fetch(url, {
      ...options,
      signal: abortController.signal,
    }),
    timeoutPromise,
  ])
    .catch((error) => {
      if (error instanceof TimeoutError) {
        abortController.abort();
      }

      throw error;
    })
    .then((response) => {
      clearTimeout(timerId);
      return response;
    }) as Promise<Response>;
}
