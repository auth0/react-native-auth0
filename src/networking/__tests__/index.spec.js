import Client from '../';
import defaults from '../telemetry';
import fetchMock from 'fetch-mock';
import {TimeoutError} from '../../utils/fetchWithTimeout';

jest.useFakeTimers();

describe('client', () => {
  const domain = 'samples.auth0.com';
  const baseUrl = `https://${domain}`;

  describe('constructor', () => {
    it('should accept only baseUrl', () => {
      const client = new Client({baseUrl});
      expect(client.baseUrl).toEqual(baseUrl);
      expect(client.telemetry).toMatchObject(defaults);
    });

    it('should accept only domain', () => {
      const client = new Client({baseUrl: domain});
      expect(client.baseUrl).toEqual(baseUrl);
      expect(client.telemetry).toMatchObject(defaults);
    });

    it('should accept only http baseUrl', () => {
      const client = new Client({baseUrl: 'http://insecure.com'});
      expect(client.baseUrl).toEqual('http://insecure.com');
      expect(client.telemetry).toMatchObject(defaults);
    });

    it('should allow to customize telemetry', () => {
      const custom = {name: 'react-native-lock', version: '1.0.0-rc.1'};
      const client = new Client({baseUrl, telemetry: custom});
      expect(client.telemetry).toMatchObject({
        ...custom,
        env: {
          'react-native-auth0': defaults.version,
        },
      });
    });

    it('should allow to specify a bearer token', () => {
      const token = 'a.bearer.token';
      const client = new Client({baseUrl, token});
      expect(client.bearer).toEqual('Bearer a.bearer.token');
    });

    it('should fail with no domain', () => {
      expect(() => new Client()).toThrowErrorMatchingSnapshot();
    });
  });

  describe('requests', () => {
    const client = new Client({
      baseUrl,
      telemetry: {name: 'react-native-auth0', version: '1.0.0'},
      token: 'a.bearer.token',
    });

    beforeEach(fetchMock.restore);

    describe('POST json', () => {
      const body = {
        string: 'value',
        number: 10,
      };
      const response = {
        body: {
          key: 'value',
        },
        headers: {
          'Content-Type': 'application/json',
        },
      };

      it('should build proper request with body', async () => {
        fetchMock.postOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await client.post('/method', body);
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it('should build proper request with no body', async () => {
        fetchMock.postOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await client.post('/method');
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it('should return json on success', async () => {
        fetchMock.postOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await expect(client.post('/method', body)).resolves.toMatchSnapshot();
      });

      it('should handle no response', async () => {
        fetchMock.postOnce('https://samples.auth0.com/method', {status: 201});
        expect.assertions(1);
        await expect(client.post('/method', body)).resolves.toMatchSnapshot();
      });

      it('should handle request error', async () => {
        fetchMock.postOnce('https://samples.auth0.com/method', {
          throws: new Error('pawned!'),
        });
        expect.assertions(1);
        await expect(client.post('/method', body)).rejects.toMatchSnapshot();
      });
    });

    describe('PATCH json', () => {
      const body = {
        string: 'value',
        number: 10,
      };
      const response = {
        body: {
          key: 'value',
        },
        headers: {
          'Content-Type': 'application/json',
        },
      };

      it('should build proper request with body', async () => {
        fetchMock.patchOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await client.patch('/method', body);
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it('should build proper request with no body', async () => {
        fetchMock.patchOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await client.patch('/method');
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it('should return json on success', async () => {
        fetchMock.patchOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await expect(client.patch('/method', body)).resolves.toMatchSnapshot();
      });

      it('should handle no response', async () => {
        fetchMock.patchOnce('https://samples.auth0.com/method', {
          status: 201,
        });
        expect.assertions(1);
        await expect(client.patch('/method', body)).resolves.toMatchSnapshot();
      });

      it('should handle request error', async () => {
        fetchMock.patchOnce('https://samples.auth0.com/method', {
          throws: new Error('pawned!'),
        });
        expect.assertions(1);
        await expect(client.patch('/method', body)).rejects.toMatchSnapshot();
      });
    });

    describe('GET json', () => {
      const query = {
        string: 'value',
        number: 10,
      };
      const response = {
        body: {
          key: 'value',
        },
        headers: {
          'Content-Type': 'application/json',
        },
      };

      it('should build proper request with query', async () => {
        fetchMock.getOnce(
          'https://samples.auth0.com/method?string=value&number=10',
          response,
        );
        expect.assertions(1);
        await client.get('/method', query);
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it('should build proper request with no query', async () => {
        fetchMock.getOnce('https://samples.auth0.com/method', response);
        expect.assertions(1);
        await client.get('/method');
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it('should return json on success', async () => {
        fetchMock.getOnce(
          'https://samples.auth0.com/method?string=value&number=10',
          response,
        );
        expect.assertions(1);
        await expect(client.get('/method', query)).resolves.toMatchSnapshot();
      });

      it('should handle no response', async () => {
        fetchMock.getOnce(
          'https://samples.auth0.com/method?string=value&number=10',
          {status: 201},
        );
        expect.assertions(1);
        await expect(client.get('/method', query)).resolves.toMatchSnapshot();
      });

      it('should handle request error', async () => {
        fetchMock.getOnce(
          'https://samples.auth0.com/method?string=value&number=10',
          {throws: new Error('pawned!')},
        );
        expect.assertions(1);
        await expect(client.get('/method', query)).rejects.toMatchSnapshot();
      });
    });
  });

  describe('url', () => {
    const client = new Client({
      baseUrl,
      telemetry: {name: 'react-native-auth0', version: '1.0.0'},
      token: 'a.bearer.token',
    });

    it('should build url with no query', () => {
      expect(client.url('/authorize', null, true)).toMatchSnapshot();
    });

    it('should build url with no query and no telemetry', () => {
      expect(client.url('/authorize', null, false)).toMatchSnapshot();
    });

    it('should build url with query', () => {
      expect(
        client.url('/authorize', {client_id: 'A_CLIENT_ID'}, true),
      ).toMatchSnapshot();
    });
  });

  describe('timeout', () => {
    const client = new Client({
      baseUrl,
      telemetry: {name: 'react-native-auth0', version: '1.0.0'},
      token: 'a.bearer.token',
      timeout: 2,
    });

    const response = {
      body: {
        key: 'value',
      },
      headers: {
        'Content-Type': 'application/json',
      },
    };

    let responseTimerId = null;

    beforeEach(() => {
      jest.useFakeTimers();
      fetchMock.restore();

      fetchMock.mock('https://samples.auth0.com/method', () => {
        return new Promise(resolve => {
          responseTimerId = setTimeout(() => {
            resolve(response);
          }, 2000);
          jest.runAllTimers();
        });
      });
    });

    afterEach(() => {
      jest.useRealTimers();
      if (responseTimerId) {
        clearTimeout(responseTimerId);
      }
    });

    it('aborts the request and throws TimeoutError if the request takes longer than specified timeout', async () => {
      await expect(client.get('/method')).rejects.toThrow(TimeoutError);

      const [_, fetchOptions] = fetchMock.lastCall();
      expect(fetchOptions.signal.aborted).toBeTruthy();
    });
  });
});
