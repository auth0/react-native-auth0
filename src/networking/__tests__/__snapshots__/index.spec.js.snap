// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client constructor should fail with no domain 1`] = `"Cannot read properties of undefined (reading 'baseUrl')"`;

exports[`client requests GET json should build proper request with no query 1`] = `
Array [
  "https://samples.auth0.com/method",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer a.bearer.token",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "signal": AbortSignal {
      Symbol(kEvents): Map {
        "abort" => Object {
          "next": Listener {
            "callback": [Function],
            "capture": false,
            "isNodeStyleListener": false,
            "listener": [Function],
            "next": undefined,
            "once": false,
            "passive": false,
            "previous": [Circular],
            "removed": false,
            "weak": false,
          },
          "size": 1,
        },
      },
      Symbol(events.maxEventTargetListeners): 10,
      Symbol(events.maxEventTargetListenersWarned): false,
      Symbol(kAborted): false,
    },
  },
]
`;

exports[`client requests GET json should build proper request with query 1`] = `
Array [
  "https://samples.auth0.com/method?string=value&number=10",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer a.bearer.token",
      "Content-Type": "application/json",
    },
    "method": "GET",
    "signal": AbortSignal {
      Symbol(kEvents): Map {
        "abort" => Object {
          "next": Listener {
            "callback": [Function],
            "capture": false,
            "isNodeStyleListener": false,
            "listener": [Function],
            "next": undefined,
            "once": false,
            "passive": false,
            "previous": [Circular],
            "removed": false,
            "weak": false,
          },
          "size": 1,
        },
      },
      Symbol(events.maxEventTargetListeners): 10,
      Symbol(events.maxEventTargetListenersWarned): false,
      Symbol(kAborted): false,
    },
  },
]
`;

exports[`client requests GET json should handle no response 1`] = `
Object {
  "headers": Headers {
    Symbol(map): Object {},
  },
  "ok": true,
  "status": 201,
  "text": "Created",
}
`;

exports[`client requests GET json should handle request error 1`] = `[Error: pawned!]`;

exports[`client requests GET json should return json on success 1`] = `
Object {
  "headers": Headers {
    Symbol(map): Object {
      "Content-Length": Array [
        "15",
      ],
      "Content-Type": Array [
        "application/json",
      ],
    },
  },
  "json": Object {
    "key": "value",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`client requests PATCH json should build proper request with body 1`] = `
Array [
  "https://samples.auth0.com/method",
  Object {
    "body": "{\\"string\\":\\"value\\",\\"number\\":10}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer a.bearer.token",
      "Content-Type": "application/json",
    },
    "method": "PATCH",
    "signal": AbortSignal {
      Symbol(kEvents): Map {
        "abort" => Object {
          "next": Listener {
            "callback": [Function],
            "capture": false,
            "isNodeStyleListener": false,
            "listener": [Function],
            "next": undefined,
            "once": false,
            "passive": false,
            "previous": [Circular],
            "removed": false,
            "weak": false,
          },
          "size": 1,
        },
      },
      Symbol(events.maxEventTargetListeners): 10,
      Symbol(events.maxEventTargetListenersWarned): false,
      Symbol(kAborted): false,
    },
  },
]
`;

exports[`client requests PATCH json should build proper request with no body 1`] = `
Array [
  "https://samples.auth0.com/method",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer a.bearer.token",
      "Content-Type": "application/json",
    },
    "method": "PATCH",
    "signal": AbortSignal {
      Symbol(kEvents): Map {
        "abort" => Object {
          "next": Listener {
            "callback": [Function],
            "capture": false,
            "isNodeStyleListener": false,
            "listener": [Function],
            "next": undefined,
            "once": false,
            "passive": false,
            "previous": [Circular],
            "removed": false,
            "weak": false,
          },
          "size": 1,
        },
      },
      Symbol(events.maxEventTargetListeners): 10,
      Symbol(events.maxEventTargetListenersWarned): false,
      Symbol(kAborted): false,
    },
  },
]
`;

exports[`client requests PATCH json should handle no response 1`] = `
Object {
  "headers": Headers {
    Symbol(map): Object {},
  },
  "ok": true,
  "status": 201,
  "text": "Created",
}
`;

exports[`client requests PATCH json should handle request error 1`] = `[Error: pawned!]`;

exports[`client requests PATCH json should return json on success 1`] = `
Object {
  "headers": Headers {
    Symbol(map): Object {
      "Content-Length": Array [
        "15",
      ],
      "Content-Type": Array [
        "application/json",
      ],
    },
  },
  "json": Object {
    "key": "value",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`client requests POST json should build proper request with body 1`] = `
Array [
  "https://samples.auth0.com/method",
  Object {
    "body": "{\\"string\\":\\"value\\",\\"number\\":10}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer a.bearer.token",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "signal": AbortSignal {
      Symbol(kEvents): Map {
        "abort" => Object {
          "next": Listener {
            "callback": [Function],
            "capture": false,
            "isNodeStyleListener": false,
            "listener": [Function],
            "next": undefined,
            "once": false,
            "passive": false,
            "previous": [Circular],
            "removed": false,
            "weak": false,
          },
          "size": 1,
        },
      },
      Symbol(events.maxEventTargetListeners): 10,
      Symbol(events.maxEventTargetListenersWarned): false,
      Symbol(kAborted): false,
    },
  },
]
`;

exports[`client requests POST json should build proper request with no body 1`] = `
Array [
  "https://samples.auth0.com/method",
  Object {
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0=",
      "Authorization": "Bearer a.bearer.token",
      "Content-Type": "application/json",
    },
    "method": "POST",
    "signal": AbortSignal {
      Symbol(kEvents): Map {
        "abort" => Object {
          "next": Listener {
            "callback": [Function],
            "capture": false,
            "isNodeStyleListener": false,
            "listener": [Function],
            "next": undefined,
            "once": false,
            "passive": false,
            "previous": [Circular],
            "removed": false,
            "weak": false,
          },
          "size": 1,
        },
      },
      Symbol(events.maxEventTargetListeners): 10,
      Symbol(events.maxEventTargetListenersWarned): false,
      Symbol(kAborted): false,
    },
  },
]
`;

exports[`client requests POST json should handle no response 1`] = `
Object {
  "headers": Headers {
    Symbol(map): Object {},
  },
  "ok": true,
  "status": 201,
  "text": "Created",
}
`;

exports[`client requests POST json should handle request error 1`] = `[Error: pawned!]`;

exports[`client requests POST json should return json on success 1`] = `
Object {
  "headers": Headers {
    Symbol(map): Object {
      "Content-Length": Array [
        "15",
      ],
      "Content-Type": Array [
        "application/json",
      ],
    },
  },
  "json": Object {
    "key": "value",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`client url should build url with no query 1`] = `"https://samples.auth0.com/authorize?auth0Client=eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0%3D"`;

exports[`client url should build url with no query and no telemetry 1`] = `"https://samples.auth0.com/authorize"`;

exports[`client url should build url with query 1`] = `"https://samples.auth0.com/authorize?client_id=A_CLIENT_ID&auth0Client=eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0%3D"`;
